sdk/cosmos/azure-cosmos/azure/cosmos/_retry_utility.py

```diff
@@ -157,99 +176,105 @@ def Execute(client, global_endpoint_manager, function, *args, **kwargs): # pylin
                 raise e_offer
 
             return result
-        except exceptions.CosmosHttpResponseError as e:
-            if request:
-                # update session token for relevant operations
-                client._UpdateSessionIfRequired(request.headers, {}, e.headers)
-            if request and _has_database_account_header(request.headers):
-                retry_policy = database_account_retry_policy
-            # Re-assign retry policy based on error code
-            elif e.status_code == StatusCodes.FORBIDDEN and e.sub_status in\
-                    [SubStatusCodes.DATABASE_ACCOUNT_NOT_FOUND, SubStatusCodes.WRITE_FORBIDDEN]:
-                retry_policy = endpointDiscovery_retry_policy
-            elif e.status_code == StatusCodes.TOO_MANY_REQUESTS:
-                retry_policy = resourceThrottle_retry_policy
-            elif (
-                e.status_code == StatusCodes.NOT_FOUND
-                and e.sub_status
-                and e.sub_status == SubStatusCodes.READ_SESSION_NOTAVAILABLE
-            ):
-                retry_policy = sessionRetry_policy
-            elif exceptions._partition_range_is_gone(e):
-                retry_policy = partition_key_range_gone_retry_policy
-            elif exceptions._container_recreate_exception(e):
-                retry_policy = container_recreate_retry_policy
-                # Before we retry if retry policy is container recreate, we need refresh the cache of the
-                # container properties and pass in the new RID in the headers.
-                client._refresh_container_properties_cache(retry_policy.container_link)
-                if e.sub_status != SubStatusCodes.COLLECTION_RID_MISMATCH and retry_policy.check_if_rid_different(
-                        retry_policy.container_link, client._container_properties_cache, retry_policy.container_rid):
-                    retry_policy.refresh_container_properties_cache = False
+        except (exceptions.CosmosHttpResponseError, ServiceRequestError, ServiceResponseError) as e:
+            # Store the last error
+            last_error = e
+
+            if timeout:
+                elapsed = time.time() - operation_start_time
+                if elapsed >= timeout:
+                    raise exceptions.CosmosClientTimeoutError(inner_exception=last_error)
```

Comment:
I still disagree that we should check the timeout here before processing the raised exception.
We should certainly check the timeout once we've determined that the error should be retried (before sleeping) - but I think otherwise - the original error should be raised.